src/symbolic_dimensions.jl

```diff
@@ -95,15 +104,122 @@ end
 expand_units(q::QuantityArray) = expand_units.(q)
 
 
-Base.copy(d::SymbolicDimensions) = SymbolicDimensions(copy(data(d)))
-Base.:(==)(l::SymbolicDimensions, r::SymbolicDimensions) = data(l) == data(r)
-Base.iszero(d::SymbolicDimensions) = iszero(data(d))
-Base.:*(l::SymbolicDimensions, r::SymbolicDimensions) = SymbolicDimensions(data(l) + data(r))
-Base.:/(l::SymbolicDimensions, r::SymbolicDimensions) = SymbolicDimensions(data(l) - data(r))
-Base.inv(d::SymbolicDimensions) = SymbolicDimensions(-data(d))
-Base.:^(l::SymbolicDimensions{R}, r::Integer) where {R} = SymbolicDimensions(data(l) * r)
-Base.:^(l::SymbolicDimensions{R}, r::Number) where {R} = SymbolicDimensions(data(l) * tryrationalize(R, r))
+Base.copy(d::SymbolicDimensions) = SymbolicDimensions(copy(getfield(d, :nzdims)), copy(getfield(d, :nzvals)))
+function Base.:(==)(l::SymbolicDimensions, r::SymbolicDimensions)
+    nzdims_l = getfield(l, :nzdims)
+    nzvals_l = getfield(l, :nzvals)
+    nzdims_r = getfield(r, :nzdims)
+    nzvals_r = getfield(r, :nzvals)
+    nl = length(nzdims_l)
+    nr = length(nzdims_r)
+    il = ir = 1
+    while il <= nl && ir <= nr
+        dim_l = nzdims_l[il]
+        dim_r = nzdims_r[ir]
+        if dim_l == dim_r
+            if nzvals_l[il] != nzvals_r[ir]
+                return false
+            end
+            il += 1
+            ir += 1
+        elseif dim_l < dim_r
+            if !iszero(nzvals_l[il])
+                return false
+            end
+            il += 1
+        else
+            if !iszero(nzvals_r[ir])
+                return false
+            end
+            ir += 1
+        end
+    end
+
+    while il <= nl
+        if !iszero(nzvals_l[il])
+            return false
+        end
+        il += 1
+    end
+
+    while ir <= nr
+        if !iszero(nzvals_r[ir])
+            return false
+        end
+        ir += 1
+    end
+
+    return true
+end
+Base.iszero(d::SymbolicDimensions) = iszero(getfield(d, :nzvals))
+
+# Defines `inv(::SymbolicDimensions)` and `^(::SymbolicDimensions, ::Number)`
+function map_dimensions(op::Function, d::SymbolicDimensions)
+    return SymbolicDimensions(getfield(d, :nzdims), map(op, getfield(d, :nzvals)))
```

Comment:
Should `nzdims` be explicitly copied here? (Since it’s mutable storage for an otherwise immutable struct)

Reply:
Good catch. Yes, probably that's safer.

Reply:
```suggestion
    return SymbolicDimensions(copy(getfield(d, :nzdims)), map(op, getfield(d, :nzvals)))
```